drivers/at_cmd/at_cmd.c

```diff
@@ -170,77 +171,54 @@ static void socket_thread_fn(void *arg1, void *arg2, void *arg3)
 			goto next;
 		}
 
-		payload_len = get_return_code(buf, &ret);
+		payload_len = get_return_code(item->data, &ret);
 
 		if (ret.state != AT_CMD_NOTIFICATION) {
 			if ((response_buf_len > 0) &&
 			    (response_buf != NULL)) {
 				if (response_buf_len > payload_len) {
-					memcpy(response_buf,
-						buf, payload_len);
+					memcpy(response_buf, item->data,
+					       payload_len);
 				} else {
 					LOG_ERR("Response buffer not large "
 						"enough");
 
 					ret.code  = -EMSGSIZE;
 				}
 
-				callback         = false;
 				response_buf_len = 0;
 				response_buf     = NULL;
+
+				goto next;
 			}
 		}
 
 		if (payload_len == 0) {
 			goto next;
 		}
 
-		if (callback) {
-			at_cmd_handler_t tmp_callback;
-
-			if (ret.state == AT_CMD_NOTIFICATION) {
-				tmp_callback = notification_handler;
-			} else {
-				tmp_callback = current_cmd_handler;
-			}
-
-			if (!tmp_callback) {
-				goto next;
-			}
-
-			struct callback_work_item *item =
-			k_malloc(sizeof(struct callback_work_item));
-
-			if (!item) {
-				LOG_DBG("Failed to allocate work item");
-
-				ret.code  = -ENOMEM;
-
-				goto next;
-			}
-
-			item->callback = tmp_callback;
-
+		if (ret.state == AT_CMD_NOTIFICATION) {
+			item->callback = notification_handler;
+		} else {
+			item->callback = current_cmd_handler;
+		}
+next:
+		/* If no callback was set, free the item.
+		 * Otherwise, work queue callback will free it.
+		 */
+		if (item->callback == NULL) {
+			k_mem_slab_free(&rsp_work_items, (void **)&item);
```

Comment:
Do we need to, couldn't we keep it and re-use it for the next iteration... Maybe that only create unnecessary complexity

Reply:
I can imagine the buffer could be reused, some extra logic would be needed when deciding whether to allocate or not. On the other I think the design is quite easy to understand, when each loop iteration is somewhat separate entity that cleans after itself. So unless there is a heavy CPU impact on allocation/freeing I'd rather keep it as it is.